NeoML/src/Dnn/Layers/FullyConnectedLayer.cpp

```diff
@@ -21,195 +21,290 @@ limitations under the License.
+		data.FaultFlag = ( data.Ptr == nullptr );
+	}
+	return data.Ptr;
+}
+
+void CFullyConnectedLayer::destroySmallMatricesMulDescs( int inputCount )
+{
+	for( int i = 0; i < TSMMD_Count_; ++i ) {
+		smallMatricesMulDescs[i].~CSmallMatricesMultiplyDescHolder();
+		new( &smallMatricesMulDescs[i] ) CSmallMatricesMultiplyDescHolder( inputCount );
+	}
 }
 
 void CFullyConnectedLayer::Reshape()
 {
 	CheckInputs();
 	CheckLayerArchitecture( GetInputCount() == GetOutputCount(),
 		"fully connected layer with different numbers of input and output" );
-	for(int i = 0; i < GetInputCount(); i++) {
-		if(Weights() == 0) {
+	for( int i = 0; i < GetInputCount(); ++i ) {
+		if( Weights() == nullptr ) {
 			// Create a weights matrix
 			CBlobDesc weightsDesc = inputDescs[i];
-			weightsDesc.SetDimSize(BD_BatchLength, 1);
-			weightsDesc.SetDimSize(BD_BatchWidth, numberOfElements);
-			weightsDesc.SetDimSize(BD_ListSize, 1);
-			Weights() = CDnnBlob::CreateBlob(MathEngine(), CT_Float, weightsDesc);
+			weightsDesc.SetDimSize( BD_BatchLength, 1 );
+			weightsDesc.SetDimSize( BD_BatchWidth, numberOfElements );
+			weightsDesc.SetDimSize( BD_ListSize, 1 );
+			Weights() = CDnnBlob::CreateBlob( MathEngine(), CT_Float, weightsDesc );
 			// Initialize
-			InitializeParamBlob(i, *Weights());
+			InitializeParamBlob( i, *Weights() );
 		} else {
 			CheckLayerArchitecture( Weights()->GetObjectCount() == numberOfElements,
 				"weights number is not equal to number of elements" );
 			CheckLayerArchitecture( Weights()->GetObjectSize() == inputDescs[i].ObjectSize(),
 				"weights size mismatch" );
 		}
 
-		if(FreeTerms() == 0) {
-			FreeTerms() = CDnnBlob::CreateVector(MathEngine(), CT_Float, numberOfElements);
+		if( FreeTerms() == nullptr ) {
+			FreeTerms() = CDnnBlob::CreateVector( MathEngine(), CT_Float, numberOfElements );
 			// Initialize
-			FreeTerms()->Fill(0);
+			FreeTerms()->Fill( 0 );
 		} else {
 			CheckLayerArchitecture( FreeTerms()->GetDataSize() == numberOfElements,
 				"free terms num is not equal to number of elements" );
 		}
 
 		// For each layer element there is a channel in the output blob
 		outputDescs[i] = inputDescs[i];
-		outputDescs[i].SetDimSize(BD_Height, 1);
-		outputDescs[i].SetDimSize(BD_Width, 1);
-		outputDescs[i].SetDimSize(BD_Depth, 1);
-		outputDescs[i].SetDimSize(BD_Channels, numberOfElements);
+		outputDescs[i].SetDimSize( BD_Height, 1 );
+		outputDescs[i].SetDimSize( BD_Width, 1 );
+		outputDescs[i].SetDimSize( BD_Depth, 1 );
+		outputDescs[i].SetDimSize( BD_Channels, numberOfElements );
 	}
+	destroySmallMatricesMulDescs( GetInputCount() );
 }
 
 void CFullyConnectedLayer::RunOnce()
 {
-	for( int i = 0; i < GetInputCount(); i++ ) {
-		CConstFloatHandle inputData = inputBlobs[i]->GetData();
-		CFloatHandle outputData = outputBlobs[i]->GetData();
-		CConstFloatHandle weightData = Weights()->GetData();
+	const int batchSize = GetInputCount();
```

Comment:
Hmmm, it looks like a misleading name, don't you think?

Reply:
fixed